README.md
# Rendertron [![Build status](https://travis-ci.org/GoogleChrome/rendertron.svg?branch=master)](https://travis-ci.org/GoogleChrome/rendertron) [![NPM rendertron package](https://img.shields.io/npm/v/rendertron.svg)](https://npmjs.org/package/rendertron)

> Rendertron is a headless Chrome rendering solution designed to render & serialise web pages on the fly.

#### :hammer: Built with [Puppeteer](https://github.com/GoogleChrome/puppeteer)
#### :cloud: Easy deployment to Google Cloud
#### :mag: Improves SEO

Rendertron is designed to enable your Progressive Web App (PWA) to serve the correct
content to any bot that doesn't render or execute JavaScript. Rendertron runs as a
standalone HTTP server. Rendertron renders requested pages using Headless Chrome,
[auto-detecting](#auto-detecting-loading-function) when your PWA has completed loading
and serializes the response back to the original request. To use Rendertron, your application
configures [middleware](#middleware) to determine whether to proxy a request to Rendertron.
Rendertron is compatible with all client side technologies, including [web components](#web-components).

**Demo endpoint**

A demo Rendertron service is available at https://render-tron.appspot.com/. It is not designed
to be used as a production endpoint. You can use it, but there are no uptime guarantees.

## Contents
- [Middleware](#middleware)
- [API](#api)
  - [Render](#render)
  - [Screenshot](#screenshot)
  - [Invalidate cache](#invalidate-cache)
- [FAQ](#faq)
  - [Query parameters](#query-parameters)
  - [Auto detecting loading function](#auto-detecting-loading-function)
  - [Rendering budget timeout](#rendering-budget-timeout)
  - [Web components](#web-components)
  - [Status codes](#status-codes)
- [Installing & deploying](#installing--deploying)
  - [Building](#building)
  - [Running locally](#running-locally)
  - [Deploying to Google Cloud Platform](#deploying-to-google-cloud-platform)
  - [Deploying using Docker](#deploying-using-docker)
  - [Config](#config)
  - [Troubleshooting](#troubleshooting)

## Middleware
Once you have the service up and running, you'll need to implement the differential serving
layer. This checks the user agent to determine whether prerendering is required.

This is a list of middleware available to use with the Rendertron service:
 * [Express.js middleware](/middleware)
 * [Firebase functions](https://github.com/justinribeiro/pwa-firebase-functions-botrender) (Community maintained)
 * [ASP.net core middleware](https://github.com/galamai/AspNetCore.Rendertron) (Community maintained)
 * [Python (Django) middleware and decorator](https://github.com/frontendr/python-rendertron) (Community maintained)

Rendertron is also compatible with [prerender.io middleware](https://prerender.io/documentation/install-middleware).
Note: the user agent lists differ there.

## API

### Render
```
GET /render/<url>
```

The `render` endpoint will render your page and serialize your page. Options are
specified as query parameters:
 * `mobile` defaults to `false`. Enable by passing `?mobile` to request the
  mobile version of your site.
 * `refreshCache`: Pass `refreshCache=true` to ignore potentially cached render results 
 and treat the request as if it is not cached yet. 
 The new render result is used to replace the previous result.  

### Screenshot
```
GET /screenshot/<url>
POST /screenshot/<url>
```

The `screenshot` endpoint can be used to verify that your page is rendering
correctly.

Both endpoints support the following query parameters:
 * `width` defaults to `1000` - specifies viewport width.
 * `height` defaults to `1000` - specifies viewport height.
 * `mobile` defaults to `false`. Enable by passing `?mobile` to request the
  mobile version of your site.
 * `timezoneId` - specifies rendering for timezone.

Additional options are available as a JSON string in the `POST` body. See
[Puppeteer documentation](https://github.com/GoogleChrome/puppeteer/blob/v1.6.0/docs/api.md#pagescreenshotoptions)
for available options. You cannot specify the `type` (defaults to `jpeg`) and
`encoding` (defaults to `binary`) parameters.

### Invalidate cache
```
GET /invalidate/<url>
```

The `invalidate` endpoint will remove cache entried for `<url>` from the configured cache (in-memory, filesystem or cloud datastore).

## FAQ

### Query parameters
When setting query parameters as part of your URL, ensure they are encoded correctly. In JS,
this would be `encodeURIComponent(myURLWithParams)`. For example to specify `page=home`:
```
https://render-tron.appspot.com/render/http://my.domain/%3Fpage%3Dhome
```

### Auto detecting loading function
The service detects when a page has loaded by looking at the page load event, ensuring there
are no outstanding network requests and that the page has had ample time to render.

### Rendering budget timeout
There is a hard limit of 10 seconds for rendering. Ensure you don't hit this budget by ensuring
your application is rendered well before the budget expires.

### Web components
Headless Chrome supports web components but shadow DOM is difficult to serialize effectively.
As such, [shady DOM](https://github.com/webcomponents/shadydom) (a lightweight shim for Shadow DOM)
is required for web components.

If you are using web components v0 (deprecated), you will need to enable Shady DOM to
render correctly. In Polymer 1.x, which uses web components v0, Shady DOM is enabled by default.
If you are using Shadow DOM, override this by setting the query parameter `dom=shady` when
directing requests to the Rendertron service.

If you are using web components v1 and either `webcomponents-lite.js` or `webcomponents-loader.js`,
set the query parameter `wc-inject-shadydom=true` when directing requests to the Rendertron
service. This renderer service will force the necessary polyfills to be loaded and enabled.

### Status codes
Status codes from the initial requested URL are preserved. If this is a 200, or 304, you can
set the HTTP status returned by the rendering service by adding a meta tag.
```html
<meta name="render:status_code" content="404" />
```

## Running locally
To install Rendertron and run it locally, first install Rendertron:
```bash
npm install -g rendertron
```

With Chrome installed on your machine run the Rendertron CLI:
```bash
rendertron
```

## Installing & deploying

### Building
Clone and install dependencies:
```bash
git clone https://github.com/GoogleChrome/rendertron.git
cd rendertron
npm install
npm run build
```

### Running locally
With a local instance of Chrome installed, you can start the server locally:
```bash
npm run start
```

### Deploying to Google Cloud Platform
```
gcloud app deploy app.yaml --project <your-project-id>
```

### Deploying using Docker
Rendertron no longer includes a Docker file. Instead, refer to
[Puppeteer documentation](https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker)
on how to deploy run headless Chrome in Docker.

### Config
When deploying the service, set configuration variables by including a `config.json` in the
root. Available configuration options:
 * `timeout` _default `10000`_ - set the timeout used to render the target page. 
 * `port` _default `3000`_ - set the port to use for running and listening the rendertron service. Note if process.env.PORT is set, it will be used instead.
 * `host` _default `0.0.0.0`_ - set the hostname to use for running and listening the rendertron service. Note if process.env.HOST is set, it will be used instead.
 * `width` _default `1000`_ - set the width (resolution) to be used for rendering the page.
 * `height` _default `1000`_ - set the height (resolution) to be used for rendering the page.
 * `reqHeaders` _default `{}`_ - set the additional HTTP headers to be sent to the target page with every request.
 * `cache` _default `null`_ - set to `datastore` to enable caching on Google Cloud using datastore _only use if deploying to google cloud_, `memory` to enable in-memory caching or `filesystem` to enable disk based caching
 * `cacheConfig` - an object array to specify caching options
 * `renderOnly` - restrict the endpoint to only service requests for certain domains. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
 * `blacklist` - reject requests to a certain domain. Specified as an array of strings. eg. `['http://dont.render.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).

#### cacheConfig
* `cacheDurationMinutes` _default `1440`_ - set an expiry time in minues, defaults to 24 hours. Set to -1 to disable cache Expiration
* `cacheMaxEntries` _default `100`_ - set the maximum number of entries stored in the selected cache method. Set to `-1` to allow unlimited caching. If using the datastore caching method, setting this value over `1000` may lead to degraded performance as the query to determine the size of the cache may be too slow. If you want to allow a larger cache in `datastore` consider setting this to `-1` and managing the the size of your datastore using a method like this [Deleting Entries in Bulk](https://cloud.google.com/datastore/docs/bulk-delete)
* `snapshotDir` _default `<your os's default tmp dir>/renderton`_ - **filesystem only** the directory the rendertron cache files will be stored in

##### Example
An example config file specifying a memory cache, with a 2 hour expiration, and a maximum of 50 entries
```javascript
{
    "cache": "memory",
    "cacheConfig": {
        "cacheDurationMinutes": 120,
        "cacheMaxEntries": 50
    }
}
```

### Troubleshooting
If you're having troubles with getting Headless Chrome to run in your
environment, refer to the
[troubleshooting guide](https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md)
for Puppeteer.


CHANGELOG.md
# Change Log

<!-- ## Unreleased -->

## [3.1.0] 2020-11-04

### Security improvements

- Prevents rendering of \*.internal URLs, patching an issue with GCE.
- Adds allow-listing option to restrict rendering to a given list of domains or URL patterns.

### Features

- Include Heroku deploy documentation and an interactive "Click to deploy" for Heroku
- Introduces an API endpoint to clear all cache
- Adds timezone support
- Adds optional forced browser restart between renders
- Adds documentation to deploy Rendertron via Docker
- Adds option to add request headers to the rendered pages in Rendertron

### Improvements

- Fixes a bug in the filesystem cache
- Fixes issue with the injected base tag
- Updates all dependencies
- Increased the required node.js version to Node.js 10+
- Updates to the FaQ

## [3.0.0] 2020-07-02

- Introduces new configuration file format for the `config.json` options (see [README.md](./README.md))
- Introduces new cache providers:
  - In-memory cache
  - File system cache
- Introduces API endpoint to invalidate cache for a URL
- Introduces a number of new configuration options
- Introduces `refreshCache` parameter to force cache update for a URL
- Relaunches browser when the browser disconnects from Puppeteer
- Now includes doctype in rendered output
- Harmonises the configuration options for caches
- Closes page after screenshot
- Fixes security issue with AppEngine deployments
- Fixes issue with specifying host and port

## [2.0.1] 2018-09-18

- Remove testing and other files from NPM package.
- Fix NPM main config.
- Improved restrictions for endpoints.
- Support for structured data by not stripping all script tags.

## [2.0.0] 2018-07-26

- Rebuilt with Puppeteer under the hood
- Rebuilt as Koa server instead of an Express server
- Rebuilt using Typescript
- Removed explicit rendering flag
- Added support for a mobile parameter
- Added more options for screenshots

## [1.1.1] 2018-01-05

- Update `debug` flag to log requested URLs to render
- Fix for renderComplete flag
- Minor bug fixes

## [1.1.0] 2017-10-27

- Initial release on NPM


src/renderer.js
/*
 * Copyright 2017 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

'use strict';

const CDP = require('chrome-remote-interface');

const LOG_WAITING_FOR_FLAG = 'Rendertron: Waiting for rendering flag';
const LOG_RENDER_COMPLETE = 'Rendertron: Rendering complete';

class Renderer {
  _loadPage(client, url, options, config) {
    /**
     * Finds any meta tags setting the status code.
     * @return {?number} status code
     */
    function getStatusCode() {
      const metaElement = document.querySelector('meta[name="render:status_code"]');
      if (!metaElement) return;
      return parseInt(metaElement.getAttribute('content')) || undefined;
    }

    /**
     * Listens for changes to the 'renderComplete' flag.
     */
    function listenToCompletionFlag() {
      Object.defineProperty(window, 'renderComplete', {
        set: function(value) {
          if (value == false) {
            console.log('Rendertron: Waiting for rendering flag');
          } else if (value == true) {
            console.log('Rendertron: Rendering complete');
          }
        }
      });
    }

    return new Promise(async(resolve, reject) => {
      let networkIdleTimeout = undefined;

      const onReject = (arg) => {
        clearTimeout(networkIdleTimeout);
        clearTimeout(maxTimeout);
        reject(arg);
      };

      const {Page, Runtime, Network, Emulation, Console} = client;

      await Promise.all([
        Page.enable(),
        Runtime.enable(),
        Console.enable(),
        Network.enable(),
      ]);

      // Inject the Shady DOM polyfill if web components v1 is used, so we can
      // serialize the page.
      // TODO(samli): This needs to change to use the non-deprecated API after Chrome 61
      //   addScriptToEvaluateOnNewDocument({source: `ShadyDOM = {force: true}`})
      const shadyFlag = options['wc-inject-shadydom'];
      if (shadyFlag == '' || !!shadyFlag) {
        // Deprecated in Chrome 61.
        Page.addScriptToEvaluateOnLoad({scriptSource: `customElements.forcePolyfill = true`});
        Page.addScriptToEvaluateOnLoad({scriptSource: `ShadyDOM = {force: true}`});
        Page.addScriptToEvaluateOnLoad({scriptSource: `ShadyCSS = {shimcssproperties: true}`});
      }

      // Add hook for completion event.
      Page.addScriptToEvaluateOnNewDocument({source: `(${listenToCompletionFlag.toString()})()`});

      Page.navigate({url: url}).catch(onReject);

      // Check that all outstanding network requests have finished loading.
      const outstandingRequests = new Map();
      let initialRequestId = undefined;

      Network.requestWillBeSent((event) => {
        if (!initialRequestId)
          initialRequestId = event.requestId;
        outstandingRequests.set(event.requestId, event);
        if (networkIdleTimeout) {
          clearTimeout(networkIdleTimeout);
          networkIdleTimeout = undefined;
          networkIdle = false;
        }
      });

      let statusCode = 200;

      Network.responseReceived((event) => {
        if (event.requestId == initialRequestId &&
          event.response.status != 0 &&
          event.response.status != 304)
          statusCode = event.response.status;
      });

      Network.loadingFinished((event) => {
        outstandingRequests.delete(event.requestId);
        if (outstandingRequests.size == 0) {
          networkIdleTimeout = setTimeout(networkNowIdle, 500);
        }
      });

      Network.loadingFailed((event) => {
        if (event.requestId == initialRequestId) {
          onReject({message: event.errorText});
          return;
        }
        outstandingRequests.delete(event.requestId);
        if (outstandingRequests.size == 0) {
          networkIdleTimeout = setTimeout(networkNowIdle, 500);
        }
      });

      // Ensure the page load event has fired.
      let pageLoadEventFired = false;
      Page.loadEventFired(() => {
        pageLoadEventFired = true;
        triggerTimeBudget();
      });

      // Called when the network has been quiet for 500ms.
      let networkIdle = false;
      let networkNowIdle = () => {
        networkIdle = true;
        triggerTimeBudget();
      };

      // Check if an explicit `renderComplete` flag is being used.
      let waitForFlag = false;
      Console.messageAdded((event) => {
        if (event.message.text === LOG_WAITING_FOR_FLAG) {
          waitForFlag = true;
        } else if (event.message.text === LOG_RENDER_COMPLETE) {
          pageReady();
        } else if (!!config['debug']) {
          console.log(`[${event.message.level}] ${event.message.text}`);
        }
      });

      let timeBudgetStarted = false;
      let triggerTimeBudget = () => {
        if (timeBudgetStarted || waitForFlag)
          return;

        // Set a virtual time budget of 5 seconds for script/rendering. Once the page is
        // idle, the virtual time budget expires immediately.
        if (networkIdle && pageLoadEventFired) {
          Emulation.setVirtualTimePolicy({policy: 'advance', budget: 5000});
          timeBudgetStarted = true;
        }
      };

      let pageReady = async function pageReadyFn() {
        // Synchronously clear timeout & pageReady() to prevent it from
        // being called again.
        pageReady = () => {};
        clearTimeout(networkIdleTimeout);
        clearTimeout(maxTimeout);

        let result = await Runtime.evaluate({expression: `(${getStatusCode.toString()})()`});

        // Original status codes which aren't 200 always return with that status code,
        // regardless of meta tags.
        if (statusCode == 200 && result.result.value)
          statusCode = result.result.value;

        resolve({status: statusCode || 200});
      };

      Emulation.virtualTimeBudgetExpired(pageReady);

      // Set a hard limit of 10 seconds.
      let maxTimeout = setTimeout(() => {
        console.log(`10 second time budget limit reached.
          Attempted rendering: ${url}
          Page load event fired: ${pageLoadEventFired}
          Outstanding network requests: ${outstandingRequests.size}`);
        pageReady();
      }, 10000);
    });
  }

  async closeConnection(id, port) {
    try {
      await CDP.Close({id: id, port: port});
    } catch (e) {
      console.log('Could not close connection');
      console.log(e);
    }
  }

  serialize(url, options, config) {
    /**
     * Executed on the page after the page has loaded. Strips script and
     * import tags to prevent further loading of resources.
     */
    function stripPage() {
      const elements = document.querySelectorAll('script, link[rel=import]');
      elements.forEach((e) => e.remove());
    }

    /**
     * Injects a <base> tag which allows other resources to load. This
     * has no effect on serialised output, but allows it to verify render quality.
     * @param {string} url - Requested URL to set as the base.
     */
    function injectBaseHref(url) {
      const actuallyLocationUrl = window.location.protocol + '//' + window.location.host + '/';
      const firstBase = document.getElementsByTagName('base').length && document.getElementsByTagName('base')[0].href;
      if (firstBase && firstBase.search(actuallyLocationUrl) == 0) {
        // It's base relative, relocate uri
        const originalHost = url.substring(0, url.indexOf('/', 8));
        const pathBase = firstBase.substring(firstBase.indexOf('/', 8));
        document.getElementsByTagName('base')[0].href = originalHost + pathBase;
      }
      const base = document.createElement('base');
      base.setAttribute('href', url);
      document.head.appendChild(base);
    }

    return new Promise(async(resolve, reject) => {
      const tab = await CDP.New({port: config.port});
      const client = await CDP({tab: tab, port: config.port});

      const {Runtime} = client;

      try {
        let renderResult = await this._loadPage(client, url, options, config);

        await Runtime.evaluate({expression: `(${stripPage.toString()})()`});
        await Runtime.evaluate({expression: `(${injectBaseHref.toString()})('${url}')`});

        let result = await Runtime.evaluate({expression: 'document.firstElementChild.outerHTML'});
        await this.closeConnection(client.target.id, config.port);
        resolve({
          status: renderResult.status,
          body: result.result.value});
      } catch (error) {
        await this.closeConnection(client.target.id, config.port);
        reject(error);
      }
    });
  }

  captureScreenshot(url, options, config) {
    return new Promise(async(resolve, reject) => {
      const tab = await CDP.New({port: config.port});
      const client = await CDP({tab: tab, port: config.port});

      const {Animation, Page, Emulation} = client;

      // Accelerate global animation timeline so that loading animations
      // are hopefully complete by the time we take the screenshot.
      Animation.setPlaybackRate({playbackRate: 1000});

      const width = Math.min(2000, parseInt(options['width']) || 1000);
      const height = Math.min(2000, parseInt(options['height']) || 1000);
      await Emulation.setDeviceMetricsOverride({width: width, height: height, mobile: true, deviceScaleFactor: 1, fitWindow: false, screenWidth: width, screenHeight: height});
      await Emulation.setVisibleSize({width: width, height: height});

      try {
        await this._loadPage(client, url, options, config);
        let {data} = await Page.captureScreenshot({format: 'jpeg', quality: 60});

        await this.closeConnection(client.target.id, config.port);
        resolve(data);
      } catch (error) {
        await this.closeConnection(client.target.id, config.port);
        reject(error);
      }
    });
  }
}

module.exports = new Renderer();
